### ISSUE TITLE
Allow a way for csv reading to restart with invalid types

### ISSUE THREAD

**Issue by nalimilan** [JuliaData/CSV.jl#200]
This bug has been reported [on Discourse](https://discourse.julialang.org/t/importing-csv-with-missing-data/10589). I've been able to simplify it to the following file, as long as the number of rows used for detection is limited to 1:
```
r,t,y
0,0,,
A,0,,
```

```julia
 CSV.read("test.csv", rows_for_type_detect=1)
ERROR: BoundsError: attempt to access 0-element Array{UInt8,1} at index [1]
Stacktrace:
 [1] parsefield(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Type{Int64}, ::CSV.Options{DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s"),Tuple{Base.Dates.DatePart{'y'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'m'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'d'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'H'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'M'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'S'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'s'}}}}, ::Int64, ::Int64, ::Base.RefValue{CSV.ParsingState}, ::CSV.##6#7) at /home/milan/.julia/CSV/src/parsefields.jl:81
 [2] macro expansion at /home/milan/.julia/DataStreams/src/query.jl:484 [inlined]
 [3] stream!(::CSV.Source{Base.AbstractIOBuffer{Array{UInt8,1}},DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s"),Tuple{Base.Dates.DatePart{'y'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'m'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'d'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'H'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'M'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'S'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'s'}}}}, ::DataStreams.Data.Query{0x01,Tuple{DataStreams.Data.QueryColumn{0x01,Union{Int64, Missings.Missing},1,1,:r,nothing,()},DataStreams.Data.QueryColumn{0x01,Union{Int64, Missings.Missing},2,2,:t,nothing,()},DataStreams.Data.QueryColumn{0x01,Missings.Missing,3,3,:y,nothing,()}},(),nothing,nothing}, ::Type{DataStreams.Data.Field}, ::DataFrames.DataFrameStream{Tuple{Array{Union{Int64, Missings.Missing},1},Array{Union{Int64, Missings.Missing},1},Array{Missings.Missing,1}}}, ::DataStreams.Data.Schema{true,Tuple{Union{Int64, Missings.Missing},Union{Int64, Missings.Missing},Missings.Missing}}, ::Int64) at /home/milan/.julia/DataStreams/src/query.jl:628
 [4] #stream!#118(::Bool, ::Array{Any,1}, ::Function, ::CSV.Source{Base.AbstractIOBuffer{Array{UInt8,1}},DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s"),Tuple{Base.Dates.DatePart{'y'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'m'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'d'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'H'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'M'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'S'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'s'}}}}, ::DataStreams.Data.Query{0x01,Tuple{DataStreams.Data.QueryColumn{0x01,Union{Int64, Missings.Missing},1,1,:r,nothing,()},DataStreams.Data.QueryColumn{0x01,Union{Int64, Missings.Missing},2,2,:t,nothing,()},DataStreams.Data.QueryColumn{0x01,Missings.Missing,3,3,:y,nothing,()}},(),nothing,nothing}, ::Type{DataFrames.DataFrame}) at /home/milan/.julia/DataStreams/src/query.jl:598
 [5] (::DataStreams.Data.#kw##stream!)(::Array{Any,1}, ::DataStreams.Data.#stream!, ::CSV.Source{Base.AbstractIOBuffer{Array{UInt8,1}},DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s"),Tuple{Base.Dates.DatePart{'y'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'m'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'d'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'H'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'M'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'S'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'s'}}}}, ::DataStreams.Data.Query{0x01,Tuple{DataStreams.Data.QueryColumn{0x01,Union{Int64, Missings.Missing},1,1,:r,nothing,()},DataStreams.Data.QueryColumn{0x01,Union{Int64, Missings.Missing},2,2,:t,nothing,()},DataStreams.Data.QueryColumn{0x01,Missings.Missing,3,3,:y,nothing,()}},(),nothing,nothing}, ::Type{DataFrames.DataFrame}) at ./<missing>:0
 [6] #stream!#116(::Bool, ::Dict{Int64,Function}, ::Function, ::Array{Any,1}, ::Array{Any,1}, ::Void, ::Void, ::Array{Any,1}, ::DataStreams.Data.#stream!, ::CSV.Source{Base.AbstractIOBuffer{Array{UInt8,1}},DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s"),Tuple{Base.Dates.DatePart{'y'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'m'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'d'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'H'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'M'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'S'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'s'}}}}, ::Type{DataFrames.DataFrame}) at /home/milan/.julia/DataStreams/src/query.jl:548
 [7] (::DataStreams.Data.#kw##stream!)(::Array{Any,1}, ::DataStreams.Data.#stream!, ::CSV.Source{Base.AbstractIOBuffer{Array{UInt8,1}},DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s"),Tuple{Base.Dates.DatePart{'y'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'m'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'d'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'H'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'M'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'S'},Base.Dates.Delim{Char,1},Base.Dates.DatePart{'s'}}}}, ::Type{DataFrames.DataFrame}) at ./<missing>:0
 [8] #read#43(::Bool, ::Dict{Int64,Function}, ::Bool, ::Array{Any,1}, ::Function, ::String, ::Type{T} where T) at /home/milan/.julia/CSV/src/Source.jl:339
 [9] (::CSV.#kw##read)(::Array{Any,1}, ::CSV.#read, ::String, ::Type{T} where T) at ./<missing>:0 (repeats 2 times)
 [10] macro expansion at ./REPL.jl:97 [inlined]
 [11] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
 ```

It looks like we cannot do better than throw an error, since a letter appears in a field which is exclusively numeric in the rows used for type detection. But the error should be more explicit, as is already the case in other situations.

Even better, we should be able to automatically restart parsing with more appropriate column types. In the real-world example (NHIS 1981 data), the first string appeared at row 18000 in a column which otherwise only contains numbers.

**Comment by quinnj** [comment]
On current master with `CSV.File`, invalid values throw a "warning" and are replaced by `missing` by default. There is an option to do `CSV.File(file; strict=true)` to throw an error instead. I've renamed this issue to track ideas/progress on even more robust ways we can allow CSV to restart parsing when invalid types are encountered.

**Comment by garborg** [comment]
I ran into this with BTS airline on-time performance data. Each month has a new csv with a different subset of the 100 or so cols having the first 10k or so of their rows missing before the first string or int is hit, at which, as far as i can tell, CSV.jl is set on a column type. Building up `types` file by file works, but I'd opt into a performance hit on this one, given the choice.

**Comment by quinnj** [comment]
Here's how the current algorithm works: up to 10K rows, the entire dataset will be parsed to determine columns types; for datasets larger than 10K rows, 100 equally spaced chunks of 100 rows (10K total) will be used to determine column types, so we should definitely be handling cases where the 1st X number of rows are blank and then there's data. The main weakness is (and will always be) very large datasets with sparse data. We could potentially make the 10K limit configurable, so if desired, a user could pass in a larger limit to use for type inference. Would that work in your case @garborg?

**Comment by garborg** [comment]
That would work, but I wouldn't want to suggest an interface that you get boxed in to. Another easy solution in my case would be to read only a subset of the columns (which aren't as sparse and which I know the types of, in any case).

**Comment by quinnj** [comment]
Implemented in #395

**Comment by quinnj** [comment]
Fixed in #395
